src/applications/model/gaming-streaming-server.h

```diff
@@ -87,38 +84,74 @@ class GamingStreamingServer : public Application
   uint32_t GetTotSentPackets (void);
 
   /**
-   * Erase the statistics of sent packets
+   * \brief return the number of total failed packets
+   * \return total number of failed packets
+   */
+  uint32_t GetTotFailedPackets (void);
+
+  /**
+   * \brief return the total bytes sent
+   * \return the total bytes sent
+   */
+  uint32_t GetTotSentBytes (void);
+
+  /**
+   * \brief Erase the statistics of sent packets
    */
   void EraseStatistics (void);
 
+  /**
+   * \brief Add parameters of new stream
+   *
+   * \param packetSize A random number generator for packet size
+   * \param interArrivalTime A random number generator for inter-arrival time
+   */
+  void AddNewTraffic (Ptr<RandomVariableStream> packetSize, Ptr<RandomVariableStream> interArrivalTime);
+
+  virtual void StartApplication (void);
+  virtual void StopApplication (void);
+
 protected:
   /**
-   * \brief Generate and send packets based on the different distribution
+   * \brief Generate and send packets based on the different random
+   * distribution for packet sizes and inter-arrival times
    *
-   * \param pktDist random variable which defines the size of generated packets
-   * \param iatDist random variable which defines inter-arrival times between the packets
+   * \param idx index of stream within the streams vector
    */
-  void SendPacket (Ptr<RandomVariableStream> pktDist, Ptr<RandomVariableStream> iatDist);
+  void Send (int idx);
 
-  uint32_t     m_sentSeq;  //!< Sequence number for packets
+  /**
+   *  Initialize the parameters of different streams
+   */
+  virtual void InitializeStreams () = 0;
+
+private:
+  /**
+   * \brief Schedule the next packet transmission
+   *
+   * \param idx index of stream
+   */
+  void ScheduleNextTx (int idx);
+
+  uint32_t     m_seq;  //!< Sequence number for packets
   uint32_t     m_totSentPackets; //!< Counter for sent packets
+  uint32_t     m_totFailedPackets; //!< Counter for failed packets
+  uint32_t     m_totSentBytes; //!< Total bytes sent so far
   Ptr<Socket>  m_socket; //!< Socket
-  EventId      m_sendEvent; //!< Event to send the next packet
   Address      m_peerAddress; //!< Remote peer address
   uint16_t     m_peerPort; //!< Remote peer port
 
-private:
-  virtual void StartApplication (void);
+  struct TrafficStream
```

Comment:
it's fairly complicated, it should implement a destructor as well

Reply:
Sorry I didn't get it, what you mean?

Reply:
it has and event and two `Ptr`s, so you should implement a destructor as discussed yesterday

Reply:
As I understood from the links you sent, we should delete the `Ptr`s or structs in the `DoDispose` method. 
But I'm not sure if it's really necessary because we don't cause any circular referencing and because the ns3 calls the DoDispose for all objects, at the end of the simulation, it removes these links.

I'm not sure anyway. You are far experienced than me in this. please let me know what do you think?
 and if you think it's necessary, is it enough to iterate over the vector of `TrafficStream` struct and delete them?